hbase/src/main/scala/akka/stream/alpakka/hbase/internal/HBaseFlowStage.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.hbase.internal
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.stream.alpakka.hbase.HTableSettings
+import akka.stream.stage._
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import org.apache.hadoop.hbase.client.Table
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success, Try }
+
+private[hbase] class HBaseFlowStage[A](settings: HTableSettings[A])
+    extends GraphStage[FlowShape[A, Future[A]]]
+    with HBaseCapabilities {
+
+  private val in = Inlet[A]("messages")
+  private val out = Outlet[Future[A]]("result")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+
+    implicit val connection = connect(settings.conf)
+
+    val table: Table = getOrCreateTable(settings.tableName, settings.columnFamilies)
+
+    new GraphStageLogic(shape) with StageLogging {
+      val awaitingConfirmation = new AtomicInteger(0)
+      @volatile var inIsClosed = false
+      var completionState: Option[Try[Unit]] = None
+
+      setHandler(out, new OutHandler {
+        override def onPull() =
+          tryPull(in)
+      })
+
+      setHandler(in, new InHandler {
+        override def onPush() = {
+          val msg = grab(in)
+          val r = Promise[A]
+
+          awaitingConfirmation.incrementAndGet()
+
+          Future {
+            table.put(settings.converter.put(msg))
+          }.map { r =>
+            if (awaitingConfirmation.decrementAndGet() == 0 && inIsClosed)
+              checkForCompletion()
```

Comment:
this looks at stage mutable state from a different thread so it is not safe, use the `asyncCallback` api to communicate with the stage from another thread instead, you can read about it in the docs here: http://doc.akka.io/docs/akka/2.4.14/scala/stream/stream-customize.html#using-asynchronous-side-channels